[Inductor] Modify TritonTemplate store_output function to support TMA stores

NikhilAPatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151775
* #151774

Summary:
The `store_output` macro -- used in Triton templates to generate triton kernel code for storing output using `tl.store` -- has been modified to support TMA based stores.

This now allows functions using TMA stores to benefit from inductor epilogue fusion. Additionally, it is now a lot easier to add TMA stores to existing kernels.

The persistent + TMA template mm template was updated to use this logic.

Test Plan:
contbuild and OSS CI

Reviewers: paulzhan

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

NikhilAPatel committed:
[Inductor] Modify TritonTemplate store_output function to support TMA stores

Summary:
The `store_output` macro -- used in Triton templates to generate triton kernel code for storing output using `tl.store` -- has been modified to support TMA based stores.

This now allows functions using TMA stores to benefit from inductor epilogue fusion. Additionally, it is now a lot easier to add TMA stores to existing kernels.

The persistent + TMA template mm template was updated to use this logic.

Test Plan:
contbuild and OSS CI

Reviewers: paulzhan

[ghstack-poisoned]

PaulZhang12 reviewed (approved):
Looks good! Are we able to write a unit test here?

NikhilAPatel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2285,35 +2285,55 @@ def decide_later():
+                tl.store_tensor_descriptor(out_desc, [{ROW_OFFSET}, {COL_OFFSET}], {value})
+                """
             )
-        elif mode is None:
```
It was done because we need lines 2311-2323 to be executed for all scenarios other than TMA (and it is important we do not execute them in TMA mode, since it will crash). We could possibly refactor to something like below:
<img width="936" alt="SCR-20250423-ntjw" src="https://github.com/user-attachments/assets/3cbc163f-c6ef-41f9-8806-7c24c0747db4" />

But then we get a bunch of warnings that `indexing` may be unbound. Let me know which way you think is cleaner.

NikhilAPatel replied:
Also looking at this again, I think we want 2314-2323 to execute in every codepath, so I'll move that out.

NikhilAPatel committed:
Update on "[Inductor] Modify TritonTemplate store_output function to support TMA stores"


Summary:
The `store_output` macro -- used in Triton templates to generate triton kernel code for storing output using `tl.store` -- has been modified to support TMA based stores.

This now allows functions using TMA stores to benefit from inductor epilogue fusion. Additionally, it is now a lot easier to add TMA stores to existing kernels.

The persistent + TMA template mm template was updated to use this logic.

Test Plan:
contbuild and OSS CI

Reviewers: paulzhan

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
